src/Castle.Core/DynamicProxy/Generators/AttributesToAlwaysReplicate.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright 2004-2021 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy.Generators
+{
+	using System;
+	using System.Collections.Generic;
+	using System.Linq;
+
+	public static class AttributesToAlwaysReplicate
+	{
+		private static readonly object lockObject = new object();
+
+		private static IList<Type> attributes;
+
+		static AttributesToAlwaysReplicate()
+		{
+			attributes = new List<Type>()
+			{
+				typeof(ParamArrayAttribute)
+			};
+		}
+
+		public static void Add(Type attribute)
+		{
+			// note: this class is made thread-safe by replacing the backing list rather than adding to it
+			lock (lockObject)
+			{
+				attributes = new List<Type>(attributes) { attribute };
+			}
+		}
+
+		public static void Add<T>()
+		{
+			Add(typeof(T));
+		}
+
+		public static bool Contains(Type attribute)
+		{
+			return attributes.Contains(attribute);
+		}
+
+		internal static bool ShouldAdd(Type attribute)
+		{
+			return attributes.Any(attr => attr == attribute);
```

Comment:
I suspect this should be `attr.IsAssignableFrom(attribute)` (like in `AttributesToAvoidReplicating`) instead of `attr == attribute`, so that the two facilities work in the same manner.

When we only had to check for `ParamArrayAttribute` (in `AttributeUtil.cs`) , using `==` was good enough and in fact more efficient, since that FCL type is `sealed` and thus cannot have subclasses. But when we allow any types to be registered, we no longer have such a guarantee.

Reply:
I have updated my branch with this change.